mart/callbacks/anomalib.py

```diff
@@ -0,0 +1,425 @@
+from kornia.color import rgb_to_hsv, hsv_to_rgb
+from kornia.geometry.transform import rotate
+from tqdm import trange
+from anomalib.metrics import create_metric_collection
+from collections import defaultdict
+
+
+__all__ = ["SemanticAdversary"]
+
+
+class SemanticAdversary(Callback):
+    """Perturbs inputs to be adversarial under semantic contraints."""
+
+    def __init__(
+        self,
+        lr: float = 5.0,
+        steps: int = 100,
+        restarts: int = 5,
+        angle_init: float = 0,
+        angle_bound: float = 90.0,
+        angle_lr_mult: float = 1,
+        hue_init: float = 0,
+        hue_bound: float = torch.pi,
+        hue_lr_mult: float = 0.02,
+        sat_init: float = 0,
+        sat_bound: float = 0.5,
+        sat_lr_mult: float = 0.02,
+        metrics: list[str] = None,
+        seed: int = 2024,
+    ):
+        super().__init__()
+
+        self.steps = steps
+        self.restart_every = steps // (restarts or 1)
+
+        self.angle_init = angle_init
+        self.angle_bound = [-angle_bound, angle_bound]
+        self.angle_lr = lr * angle_lr_mult
+
+        self.hue_init = hue_init
+        self.hue_bound = [-hue_bound, hue_bound]
+        self.hue_lr = lr * hue_lr_mult
+
+        self.sat_init = sat_init
+        self.sat_bound = [-sat_bound, sat_bound]
+        self.sat_lr = lr * sat_lr_mult
+
+        self.metrics = metrics
+        if self.metrics is None:
+            self.metrics = ["AUROC"]
+
+        torch.manual_seed(seed)
+
+    def on_test_epoch_start(self, trainer, pl_module):
+        self.best = defaultdict(list)
+        self.history = defaultdict(list)
+
+    @torch.inference_mode(False)
+    def on_test_batch_start(
+        self, trainer, pl_module, batch, batch_idx, dataloader_idx=0
+    ):
+        device = pl_module.device
+
+        # Create optimization variables for angle, hue, and saturation
+        # FIXME: We should make this a proper nn.Module
+        batch_size = batch["image"].shape[0]
+        angle = torch.tensor(
+            [self.angle_init] * batch_size,
+            device=device,
+            dtype=torch.float32,
+            requires_grad=True,
+        )
+        hue = torch.full_like(angle, self.hue_init, requires_grad=True)
+        sat = torch.full_like(angle, self.sat_init, requires_grad=True)
+
+        # Metrics and parameters to save
+        metrics = create_metric_collection(self.metrics, prefix="p")
+
+        # FIXME: It would be nice to extract these from trainer.datamodule.test_data.transform
+        image_mean = torch.tensor([0.48145466, 0.4578275, 0.40821073], device=device)
+        image_std = torch.tensor([0.26862954, 0.26130258, 0.27577711], device=device)
+
+        # FIXME: Ideally we would clone the trainer and run fit instead of creating our own optimization loop
+        # Run optimization
+        best_batch = None
+        batch_history = defaultdict(list)
+
+        for step in (pbar := trange(self.steps, desc="Attack", position=1)):
```

Comment:
This `:=` is a bit of an exotic syntax, doesn't seem to appear anywhere else in the MART repo

Reply:
There's this: https://github.com/IntelLabs/MART/blob/36609c7f62412a95e401d75f5c6a34f1a2c926d2/mart/callbacks/progress_bar.py#L16 - not sure how helpful it is to re-use

Reply:
Ah you should learn about the walrus (:= a Python 3.8 feature)! It's super useful in some circumstances.

I think using ProgressBar will be better once this code becomes MART native.

Reply:
Good article here about the walrus operator: https://realpython.com/python-walrus-operator/